### PR TITLE
[main] feat(moe): Support packed sequence for gated delta net (GDN)

### LINKED ISSUE #4043: [Question] Enable GDN Packed Sequence Support for Context Parallelism in Qwen 3.5 series
### Core
Finetuning **Qwen3.5  long context** commonly needs **Context Parallelism (CP).**

The **Megatron-LM dev branch** has supported CP already. But **CP needs PackedSequence** and the dev branch has not support it yet. 

- So **No GDN PackedSequence**, No **Qwen3.5 training**.

- So for **training Qwen3.5**, we need **GDN PackedSequence support**.

And I'm not sure whether **you will have a plan for GDN PackedSequence**?

### Details
Megatron Repo Code
[megatron/core/ssm/gated_delta_net.py](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/ssm/gated_delta_net.py#L524)
``` python
if packed_seq_params is not None:
      # TODO: support packed sequence
      raise NotImplementedError("GDN does not support packed sequence for now.")
```

A related issue has said:
https://github.com/NVIDIA/Megatron-LM/issues/3881

Image

Verl mcore code: [verl/models/mcore/util.py](https://github.com/verl-project/verl/blob/main/verl/models/mcore/util.py)
``` python
cp_size = mpu.get_context_parallel_world_size()
assert cp_size == 1, "Context parallel size without seq_pack is not supported"
```

### PR DESCRIPTION
**Pull request by yuzhongw-nvidia**
# What does this PR do ?

Support packed sequence for gated delta net (GDN).

PR for dev: #2644 , #4230
Closes: https://github.com/NVIDIA/Megatron-LM/issues/4043, https://github.com/NVIDIA/Megatron-LM/issues/3798

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/ssm/gated_delta_net.py [hunk 1]
```diff
@@ -297,30 +297,79 @@ def forward(
...
+                        self.qk_dim_local_tp,
+                        self.qk_dim_local_tp,
+                        self.v_dim_local_tp,
+                        self.v_dim_local_tp,
+                        self.num_value_heads // self.tp_size,
+                        self.num_value_heads // self.tp_size,
+                    ],
+                )
+                outputs.append(qkvzba_i)
+            qkvzba = torch.cat(outputs, dim=0)
```
[the remark this answers is not in the document]
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/gated_delta_net.py:359]: Will not apply this change for now. Although it may reduce the launch overhead, it also introduces much communication overhead. Besides, we're developing another CP style, i.e., All Gather CP, which works much better for GDN and GDN + packed sequence, which do not need to unpack and re-pack anymore.

File: megatron/core/ssm/gated_delta_net.py [hunk 2]
```diff
@@ -297,30 +297,79 @@ def forward(
...
+                f"but got {cu_seqlens_q=} and {cu_seqlens_kv=}"
+            )
+            num_packed_seqs = cu_seqlens_q.shape[0] - 1
+            assert num_packed_seqs > 0, (
+                "Number of packed sequences must be greater than 0, "
+                f"but got {cu_seqlens_q=} and {cu_seqlens_kv=}"
+            )
+        else:
+            cu_seqlens_q = None
+            cu_seqlens_kv = None
```
[the remark this answers is not in the document]
Comment (yuzhongw-nvidia) [reply megatron/core/ssm/gated_delta_net.py:332]: Not required.

**Review by santhnm2 — APPROVED** [review]

**Review by ericharper — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
